pkg/volume/util/atomic_writer.go

```diff
@@ -0,0 +1,300 @@
+		return nil
+	}
+
+	tsDir, err := w.newTimestampDir()
+	if err != nil {
+		return err
+	}
+
+	err = w.writePayloadToDir(payload, tsDir)
+	if err != nil {
+		glog.Errorf("Unable to write payload to dir %v: %v", tsDir, err)
+	}
+
+	err = w.createPayloadSymlinksInTargetDir(payload)
+	if err != nil {
+		glog.Errorf("Unable to create payload symlinks in %v: %v", w.TargetDir, err)
+		return err
+	}
+
+	_, tsDirName := filepath.Split(tsDir)
+	currentStateDirPath := path.Join(w.TargetDir, w.currentStateDirName())
+	oldTsDir, err := os.Readlink(currentStateDirPath)
+	if err != nil && !os.IsNotExist(err) {
+		glog.Errorf("Unable to read link for current state directory: %v", err)
+	}
+
+	updateStateDirPath := path.Join(w.TargetDir, w.updateStateDirName())
+	err = os.Symlink(tsDirName, updateStateDirPath)
+	if err != nil {
+		os.RemoveAll(tsDir)
+		glog.Errorf("Unable to create symolic link: %s", err.Error())
+		return err
+	}
+
+	err = os.Rename(updateStateDirPath, currentStateDirPath)
+	if err != nil {
+		os.Remove(updateStateDirPath)
+		os.RemoveAll(tsDir)
+		glog.Errorf("Unable to rename symbolic link %v: %v", updateStateDirPath, err)
+		return err
+	}
+	if len(oldTsDir) > 0 {
+		err = os.RemoveAll(path.Join(w.TargetDir, oldTsDir))
+		if err != nil {
+			glog.Errorf("Unable to remove directory %v: %v", oldTsDir, err)
+		}
+	}
+
+	// TODO: determine error handling around sentinel file
+	return w.touchSentinelFile()
+}
+
+// shouldWritePayload returns whether the payload must be written to disk
+func (w *AtomicWriter) shouldWritePayload(payload map[string][]byte) (bool, error) {
+	for projectionPath, content := range payload {
+		shouldWrite, err := w.shouldWriteFile(path.Join(w.TargetDir, path.Clean(projectionPath)), content)
+		if err != nil {
+			return false, err
+		}
+
+		if shouldWrite {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// shouldWriteFile returns whether a new version of a file should be written to disk.
+func (w *AtomicWriter) shouldWriteFile(path string, content []byte) (bool, error) {
+	_, err := os.Lstat(path)
+	if err != nil && os.IsNotExist(err) {
+		return true, nil
+	}
+
+	contentOnFs, err := ioutil.ReadFile(path)
+	if err != nil {
+		return false, err
+	}
+
+	return (bytes.Compare(content, contentOnFs) != 0), nil
+}
+
+// newTimestampDir creates a new timestamped directory
+func (w *AtomicWriter) newTimestampDir() (string, error) {
+	tsDir, err := ioutil.TempDir(w.TargetDir, fmt.Sprintf("..%v", ts()))
+	if err != nil {
+		glog.Errorf("Unable to create new temp directory: %v", err)
+		return "", err
+	}
+
+	return tsDir, nil
+}
+
+// writePayloadToDir writes the given payload to the given directory.  The
+// directory must exist.
+func (w *AtomicWriter) writePayloadToDir(payload map[string][]byte, dir string) error {
+	allErrs := []error{}
+	for projectionPath, content := range payload {
+		fullPath := path.Join(dir, path.Clean(projectionPath))
```

Comment:
do we need to sanitize the payload paths to make sure that they don't start with a series of `../`es?